libs/execution_context/ebpf_native.c

```diff
@@ -0,0 +1,1248 @@
+{
+    UNREFERENCED_PARAMETER(context);
+
+    // 1. Find the entry in the hash table using "client_id"
+    // 2. Release the "attach" reference on the native module.
+    // 3. Return EBPF_PENDING
+    ebpf_result_t result = EBPF_PENDING;
+    ebpf_native_t** existing_native_module = NULL;
+    ebpf_native_t* native_module = NULL;
+    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_native_client_table_lock);
+    bool lock_acquired = true;
+    if (ebpf_hash_table_find(
+            _ebpf_native_client_table, (const uint8_t*)client_id, (uint8_t**)&existing_native_module) != EBPF_SUCCESS) {
+        result = EBPF_SUCCESS;
+        goto Done;
+    }
+    ebpf_lock_unlock(&_ebpf_native_client_table_lock, state);
+    lock_acquired = false;
+    native_module = *existing_native_module;
+    state = ebpf_lock_lock(&native_module->lock);
+    ebpf_assert(native_module->detaching == false);
+    native_module->detaching = true;
+    ebpf_lock_unlock(&native_module->lock, state);
+    ebpf_native_release_reference(native_module);
+
+Done:
+    if (lock_acquired) {
+        ebpf_lock_unlock(&_ebpf_native_client_table_lock, state);
+    }
+    return result;
+}
+
+ebpf_result_t
+ebpf_native_initiate()
+{
+    ebpf_result_t return_value;
+    GUID provider_module_id;
+    bool hash_table_created = false;
+
+    ebpf_lock_create(&_ebpf_native_client_table_lock);
+
+    return_value = ebpf_hash_table_create(
+        &_ebpf_native_client_table,
+        ebpf_allocate,
+        ebpf_free,
+        sizeof(GUID),
+        sizeof(ebpf_native_t*),
+        EBPF_CLIENT_TABLE_BUCKET_COUNT,
+        NULL);
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+    hash_table_created = true;
+
+    return_value = ebpf_guid_create(&provider_module_id);
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    return_value = ebpf_provider_load(
+        &_ebpf_native_provider,
+        &_ebpf_native_npi_id,
+        &provider_module_id,
+        NULL,
+        NULL,
+        NULL,
+        NULL,
+        _ebpf_native_client_attach_callback,
+        _ebpf_native_client_detach_callback);
+
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+Done:
+    if (return_value != EBPF_SUCCESS) {
+        if (hash_table_created) {
+            ebpf_hash_table_destroy(_ebpf_native_client_table);
+        }
+        ebpf_lock_destroy(&_ebpf_native_client_table_lock);
+    }
+
+    return return_value;
+}
+
+static ebpf_native_map_t*
+_ebpf_native_get_next_map_to_create(_In_ ebpf_native_map_t* maps, size_t map_count)
+{
+    for (uint32_t i = 0; i < map_count; i++) {
+        ebpf_native_map_t* map = &maps[i];
+        if (map->handle != ebpf_handle_invalid) {
+            // Already created.
+            continue;
+        }
+        if (map->entry->definition.type != BPF_MAP_TYPE_ARRAY_OF_MAPS &&
+            map->entry->definition.type != BPF_MAP_TYPE_HASH_OF_MAPS) {
```

Comment:
```suggestion
        if (!_ebpf_native_is_map_in_map(map)) {
```
and move up the definition from line 508 so it can be used here